rest/route/host.go

```diff
@@ -16,13 +16,106 @@ import (
 	"github.com/pkg/errors"
 )
 
+////////////////////////////////////////////////////////////////////////
+//
+// PATCH /rest/v2/hosts
+
+type hostsChangeStatusesHandler struct {
+	HostToStatus map[string]map[string]string
+	sc           data.Connector
+}
+
+func makeChangeHostsStatuses(sc data.Connector) gimlet.RouteHandler {
+	return &hostsChangeStatusesHandler{
+		sc: sc,
+	}
+}
+
+func (h *hostsChangeStatusesHandler) Factory() gimlet.RouteHandler {
+	return &hostsChangeStatusesHandler{
+		sc: h.sc,
+	}
+}
+
+func (h *hostsChangeStatusesHandler) Parse(ctx context.Context, r *http.Request) error {
+	body := util.NewRequestReader(r)
+	defer body.Close()
+	if err := util.ReadJSONInto(body, &h.HostToStatus); err != nil {
+		return errors.Wrap(err, "Argument read error")
+	}
+
+	if len(h.HostToStatus) == 0 {
+		return fmt.Errorf("Missing host id and status")
+	}
+	for host, json := range h.HostToStatus {
+		if !util.StringSliceContains(evergreen.ValidUserSetStatus, json["status"]) {
+			return fmt.Errorf("Invalid host status '%s' for host '%s'", json["status"], host)
+		}
+	}
+
+	return nil
+}
+
+func (h *hostsChangeStatusesHandler) Run(ctx context.Context) gimlet.Responder {
+	user := MustHaveUser(ctx)
+	for id := range h.HostToStatus {
+		foundHost, err := h.sc.FindHostByIdWithOwner(id, user)
+		if err != nil {
+			return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "Database error for find() by distro id '%s'", id))
+		}
+		if foundHost.Status == evergreen.HostTerminated {
+			return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+				StatusCode: http.StatusBadRequest,
+				Message:    fmt.Sprintf("Host '%s' is terminated; its status cannot be changed", foundHost.Id),
+			})
+		}
+	}
+
+	resp := gimlet.NewResponseBuilder()
+	if err := resp.SetFormat(gimlet.JSON); err != nil {
+		return gimlet.MakeJSONErrorResponder(err)
+	}
+
+	for id, status := range h.HostToStatus {
+		foundHost, err := h.sc.FindHostByIdWithOwner(id, user)
+		if err != nil {
+			return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "Database error for find() by distro id '%s'", id))
+		}
+		if status["status"] == evergreen.HostTerminated {
```

Comment:
if status["status"] is the empty string does SEtHostStatus handle that correctly?

Reply:
`SetHostStatus` accepts an empty string.

However, `Parse` returns an error if `!util.StringSliceContains(evergreen.ValidUserSetStatus, json["status"])`

Should I add the same validation in `Run` too?

Reply:
no need to duplicate validation.